modules/cache/context.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package cache
+
+import "context"
+
+type cacheContext struct {
+	ctx  context.Context
+	Data map[any]map[any]any
+}
+
+type cacheContextKey struct{}
+
+func WithCacheContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, &cacheContextKey{}, &cacheContext{
+		ctx:  ctx,
+		Data: make(map[any]map[any]any),
```

Comment:
I have a feeling that making this an unlimited map instead of something limited in size will bite us as more and more things use to use the cache.

Whilst LRUs have an overhead it might be a better idea to make this a generous LRU. 

I guess it might be sensible to provide a few implementations of a context cache (LRU/hard limit map/no limit map) here, AND/OR to add some trace logging/pprof metrics that counts the number of entities that gets stored in these maps. 

---

As an aside the complicated changes within models/settings do show the failure of the current design of our db services - these probably need to be migrated to service interfaces which would allow for simpler cache wrapping interfaces.

Reply:
For a request, I cannot imagine it's an unlimited map. And in fact, all the values are pointers, only keys are in this map. When a request finished, the map and related objects will be GC.